### PR TITLE
Implement type checking for callbacks

### LINKED ISSUE #26: Add support for typed callbacks
Implement typing support for callbacks.

Some example code to use as start:

```python
from typing import TypeVar, Generic, Callable, Type, overload

from oop_ext.foundation.callback import Callback


T1 = TypeVar("T1")
T2 = TypeVar("T2")

class Callback1(Generic[T1]):

    def Register(self, func: Callable[[T1], None]):  ...

    def __call__(self, t1: T1) -> None:...

class Callback2(Generic[T1, T2]):

    def Register(self, func: Callable[[T1, T2], None]):  ...

    def __call__(self, t1: T1, t2: T2) -> None:...



@overload
def TypedCallback(t1: Type[T1]) -> Callback1[T2]:
    ...

@overload
def TypedCallback(t1: Type[T1], t2: Type[T2]) -> Callback2[T1, T2]:
    ...

def TypedCallback(*types):
    return Callback()

```

In action:

```python
def on_callback(x: float, i: int) -> None:
    print(x, i)

callback = TypedCallback(str, int)
callback.Register(on_callback)
callback([], 10)
```

This gives two errors as expected:

```
foo.py:39: error: Argument 1 to "Register" of "Callback2" has incompatible type "Callable[[float, int], None]"; expected "Callable[[str, int], None]"
foo.py:40: error: Argument 1 to "__call__" of "Callback2" has incompatible type "List[<nothing>]"; expected "str"
Found 2 errors in 1 file (checked 1 source file)
```

### PR DESCRIPTION
**Pull request by nicoddemus**
Fix #26

cc @GabrielMarcilio

### CODE REVIEW

File: src/oop_ext/conftest.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+from typing import Any, TYPE_CHECKING
+
+import pytest
+
+if TYPE_CHECKING:
+    from ._type_checker_fixture import TypeCheckerFixture
+
+
+@pytest.fixture
+def type_checker(request: Any, tmp_path: Path) -> "TypeCheckerFixture":
```
Comment (williamjamir) [inline src/oop_ext/conftest.py:13]: The usage of `Any` on request is on purpose? 
Looks like pytest 6.2.0 exports pytest.FixtureRequest
```suggestion
def type_checker(request: pytest.FixtureRequest, tmp_path: Path) -> "TypeCheckerFixture":
```

**Review by williamjamir — APPROVED** [review]

File: src/oop_ext/conftest.py (hunk 1 above)
Comment (nicoddemus) [reply src/oop_ext/conftest.py]: Indeed, thanks!

File: docs/callbacks.rst [hunk 2]
```diff
@@ -0,0 +1,51 @@
+Callbacks
+=========
+
+.. automodule:: oop_ext.foundation.callback._callback
+    :noindex:
+
+
+Type Checking
```
Comment (igortg) [inline docs/callbacks.rst:8]: Nice docs

File: src/oop_ext/_type_checker_fixture.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+    @property
+    def report_lines(self) -> List[str]:
+        lines = [x.strip() for x in self.normal_report.split("\n") if x.strip()]
+        # Drop last line (summary).
+        return lines[:-1]
+
+
+@attr.s(auto_attribs=True)
+class TypeCheckerFixture:
```
Comment (nicoddemus) [inline src/oop_ext/_type_checker_fixture.py:66]: Hey @bluetech what do you think of this? Overkill, or worthwhile turning into a plugin?

Comment (igortg) [inline src/oop_ext/_type_checker_fixture.py:66]: 🧐

**Review by igortg — APPROVED** [review]

File: src/oop_ext/_type_checker_fixture.py (hunk 3 above)
Comment (bluetech) [reply src/oop_ext/_type_checker_fixture.py:66]: I think a pytest plugin is definitely useful for projects which want to verify their typing in a precise way.

I am aware of existing plugin https://github.com/davidfritzsche/pytest-mypy-testing, haven't checked its implementation.

Another comparison is to mypy's own tests, which use files like this: https://github.com/python/mypy/blob/master/test-data/unit/check-unreachable-code.test. They use pytest but I haven't checked how they run these.

BTW (unrelated), in the future you might be able to replace the 5 Callback types one just one using [Variadic Generics](https://www.python.org/dev/peps/pep-0646/), if it get accepted and implemented.

Comment (nicoddemus) [reply src/oop_ext/_type_checker_fixture.py:66]: > I am aware of existing plugin davidfritzsche/pytest-mypy-testing, haven't checked its implementation.

> Another comparison is to mypy's own tests, which use files like this: https://github.com/python/mypy/blob/master/test-data/unit/check-unreachable-code.test. They use pytest but I haven't checked how they run these.

Cool, I did a cursory search and missed those, I will check them out, thanks!

> BTW (unrelated), in the future you might be able to replace the 5 Callback types one just one using Variadic Generics, if it get accepted and implemented.

Yes, I'm aware of that, let's hope that gets in! 🤞 

Thanks for the feedback. 👍

**Pull request merged**